numpy/core/src/multiarray/arrayobject.c

```diff
@@ -469,17 +483,15 @@ array_dealloc(PyArrayObject *self)
         int retval;
         if (PyArray_FLAGS(self) & NPY_ARRAY_WRITEBACKIFCOPY)
         {
-            char * msg = "WRITEBACKIFCOPY requires a call to "
-                "PyArray_ResolveWritebackIfCopy or "
-                "PyArray_DiscardWritebackIfCopy before array_dealloc is "
-                "called.";
-            /* 2017-Nov-10 1.14 */
-            if (DEPRECATE(msg) < 0) {
-                /* dealloc cannot raise an error, best effort try to write
-                   to stderr and clear the error
-                */
-                PyErr_WriteUnraisable((PyObject *)&PyArray_Type);
-            }
+            char const * msg = "ERROR: WRITEBACKIFCOPY detected in array_dealloc. "
```

Comment:
Why not raise a `RuntimeError`, rather than a warning?

Reply:
a true exception cannot be raised in ``dealloc``, the best we can do is a warning, and if that fails print to ``stderr``

Reply:
It can't actually propagate it, but you can still call `PyError_SetString` followed by `WriteUnraisable`

Reply:
I think for the current deprecation cycle this should remain a warning. At some future point we should try as much as possible to raise an error, but not yet

Reply:
I see no reason not to raise an error here. It doesn't break compatibility in any way, and it changes the output from
```
RuntimeWarning: ERROR: ...
```
to
```
RuntimeError: ...
```

Reply:
Isn't there an advantage to being able to filter the warning, since the code will still be backward-compatible? If we raise an error, how can users filter the output?

Reply:
That's true, warnings allow the user to filter it - I'd argue it's not a compatibility break to print an unfilterable warning, but it might be still a nice thing to do.